### DISCUSSION TITLE
Lava 0.1.0

### DISCUSSION THREAD

**Discussion by mgkwill** [lava-nc/lava#5] [Announcements]
# Release 0.1.0

This first release of Lava introduces its high-level, hardware-agnostic API for developing algorithms of distributed, parallel, and asynchronous processes that communicate with each other via message-passing over channels with each other. The API is released together with the Lava compiler and runtime which together form the Magma layer of the Lava software framework. 

Our initial version of Magma allows you to familiarize yourself with the Lava user interface and to build first algorithms in Python that can be executed on CPU without requiring access to physical or cloud based Loihi resources.

# New Features and Improvements

* New Lava API to build networks of interacting Lava processes 
* New Lava Compiler to map Lava processes to executable Python code for CPU execution (support for Intel Loihi will follow)
* New Lava Runtime to execute Lave processes
* A range of fundamental tutorials illustrating the basic concepts of Lava

# Bug Fixes and Other Changes

* This is the first release of Lava. No bug fixes or other changes

# Thanks to our Contributors

@GaboFGuerra, @joyeshmishra, @PhilippPlank, @drager-intel, @mathisrichter, @srrisbud, @ysingh7, @phstratmann, @mgkwill, @awintel 
  
# Breaking Changes

* This is the first release of Lava. No breaking or other changes. 

# Known Issues

* No support for Intel Loihi yet
* Multiprocessing and channel-based communication not very performant yet
* Virtual ports for reshaping and concatenation are not supported yet
* No support for direct memory access via RefPorts yet
* Connectivity from one to many or from many to one port not supported yet
* No support for live state monitoring yet
* Still limited API documentation

<hr /><em>This discussion was created from the release <a href='https://github.com/lava-nc/lava/releases/tag/v0.1.0'>Lava 0.1.0</a>.</em>

**Comment by zhangs22** [comment]
Hi, I wonder what kinds of learning methods will be supported.

**Reply by mgkwill** [reply]
Hi @zhangs22 can you be more specific with your question?

Do you mean deep learning methods?

**Reply by zhangs22** [reply]
Thanks for your reply. I mean the learning methods such as STDP or one-pass rank-order learning for the eSNN.

**Comment by awintel** [comment]
Hi @zhangs22,
Lava in general will be completely flexible allowing you to implement any learning rule you want.
For execution on Loihi 1 or 2, STDP-inspired learning rules have to be local however, there as some differences between Loihi 1 and 2:

1. Loihi 1 supports global reward signal where the reward signal can be assigned by shared pre-synaptic axon. That means synaptic learning rules of synapses of the same axon have access to the same global reward signal. However, we found this not to be very useful. Furthermore, Loihi 1 only supports 2-factor learning rules. 3-factor learning rules are possible but require hacks involving the embedded processors of Loihi 1.
2. Loihi 2 supports 3-factor learning rules natively without hacks. In particular, the post synaptic signals available in a synaptic learning rule are not limited to passive exponential traces anymore but can be more general function of post-synaptic neuron state variables.

We do not have any examples right now. But as soon as we release them we will provide more documentation around that.

**Reply by zhangs22** [reply]
Great, thanks for your reply.

**Reply by russelljjarvis** [reply]
I would also like to use STDP with Lava/Magma with the intention of targeting Loihi hardware, and the Python Loihi simulator.

Is Lava interoperable with STDP code expressed with nxSDK or SNIPs?

Also, I wonder can neurons available via Lava/DL/Slayer be compiled to target the Loihi hardware? I am asking about that because I would like to use Lava to make network models that have both STDP and synaptic delays using Loihi. Slayer has delays at least. 

Also if I wanted to try to build STDP and synaptic delay myself, would I start with something that inherits from the AbstractProcess class?

Thanks for your help.

**Reply by awintel** [reply]
We are planning on reusing the LearningRule parser and learning rule micro code generator from the NxSDK era in Lava.
However, SNIPs from the NxSDK era won't be compatible with Lava anymore. At least not totally. We have completely overhauled and simplified how users program embedded processors. You can still use C but we added a new interface to communicate with other processes via ports with send/receive functions for message passing. So you don't have to interact with low level registers anymore.

The latest lava-dl release included the NetX package that allows you to convert SLAYER models into Lava process graphs that you can run on any backend.

We are working on enabling on-chip learning and expect to be releasing this in an upcoming release later in Q2.

**Reply by russelljjarvis** [reply]
Hi @awintel,

Thanks for your reply. These issues are relevant to me as I have been implementing [Potjans](https://www.opensourcebrain.org/projects/potjansdiesmann2014) biological model using Lava.  Do you think I will be able to discuss these issues in the upcoming Loihi/Lava workshop?

**Reply by awintel** [reply]
There will be two Lava WG group sessions throughout the week on Wednesday and Friday. That would be a chance to talk about such things.  Might be of broader interest...
There will also be different slack channels to discus topics asynchronously with one of the team members.

**Reply by russelljjarvis** [reply]
Oh cool thanks.
